statsd/buffered_metric_context.go

```diff
@@ -51,29 +53,37 @@ func (bc *bufferedMetricContexts) flush(metrics []metric) []metric {
 }
 
 func (bc *bufferedMetricContexts) sample(name string, value float64, tags []string, rate float64) error {
-	if !shouldSample(rate, bc.random, &bc.randomLock) {
-		return nil
-	}
+	keepingSample := shouldSample(rate, bc.random, &bc.randomLock)
 
 	context, stringTags := getContextAndTags(name, tags)
+	var v *bufferedMetric = nil
 
 	bc.mutex.RLock()
-	if v, found := bc.values[context]; found {
-		v.sample(value)
-		bc.mutex.RUnlock()
-		return nil
-	}
+	v, _ = bc.values[context]
 	bc.mutex.RUnlock()
 
+	// Create it if it wasn't found
```

Comment:
Just to avoid the map lookups + bc.mutex locks/unlocks, I would consider an early return here, I don't think it is adding any confusion to the simple flow of this function.

```
// sampler gonna sampling 🎵 
if !keepingSample {
  v.skipSample()
  return nil
}
```

Not a strong opinion.

Reply:
We cannot `v.skipSample` if we are not checking (and  sometimes creating) the metric first :/